x/mongo/driver/topology/connection.go

```diff
@@ -374,6 +374,17 @@ func (c *connection) readWireMessage(ctx context.Context, dst []byte) ([]byte, e
 		}
 	}
 
+	// In the case of an isMaster response where MaxMessageSize has not yet been set, use a hard-coded
+	// 48000 bytes instead.
+	maxMessageSize := c.desc.MaxMessageSize
+	if maxMessageSize == 0 {
+		maxMessageSize = 48000
+	}
+	if uint32(len(dst)) > maxMessageSize {
```

Comment:
Per DRIVERS-276, we should be checking the length reported by the header before we attempt to read the rest of the message. This allows us to detect invalid messages before we attempt to allocate that many bytes. See the definition of the `MsgHeader` struct in https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#op_msg-1 for an idea of what this looks like.

The actual fix will need to go in the private `c.read()` function which first reads 4 bytes to get the reported message length and then reads that many bytes from the socket.

Reply:
Good catch! Moved the logic to the private `c.read()` function and examined the `size` derived from the first 4 bytes of the message.